pkg/devfile/adapters/kubernetes/component/podwatcher.go

```diff
@@ -0,0 +1,441 @@
+package component
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/openshift/odo/pkg/machineoutput"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog"
+)
+
+const (
+	// KubernetesResourceFailureInterval is the time between attempts to acquire needed k8s resources
+	KubernetesResourceFailureInterval = time.Duration(5) * time.Second
+)
+
+// podWatcher is responsible for watching for changes to odo-managed Pods, and reporting those changes to the console, as used by the status command
+type podWatcher struct {
+	adapter              *Adapter
+	statusReconcilerChan chan statusReconcilerChannelEntry
+}
+
+// StartContainerStatusWatch outputs Kubernetes pod/container status changes to the console, as used by the status command
+func (a Adapter) StartContainerStatusWatch() {
+
+	pw := newPodWatcher(&a)
+	pw.startPodWatcher()
+
+}
+
+func newPodWatcher(adapter *Adapter) *podWatcher {
+	return &podWatcher{
+		adapter:              adapter,
+		statusReconcilerChan: createStatusReconciler(adapter),
+	}
+}
+
+func (pw *podWatcher) startPodWatcher() {
+	pw.startWatchThread(pw.adapter)
+}
+
+// statusReconcilerChannelEntry is the gochannel message sent from the Watcher to the status reconciler
+type statusReconcilerChannelEntry struct {
+
+	// If isCompleteListOfPods is true: a list of all component pods in the workspace
+	// If isCompleteListOfPods is false: a single component pod in the workspace
+	pods []*corev1.Pod
+
+	err error
+
+	// isCompleteListOfPods is true if the pods came from getLatestContainerStatus(), false otherwise
+	isCompleteListOfPods bool
+
+	// isDeleteEventFromWatch is true if watch.Deleted event from the watch, false otherwise
+	isDeleteEventFromWatch bool
+
+	// watchThreadRestarted is true if the watch thread died (for example, due to losing network connection) and had to be reestablished
+	isWatchThreadRestarted bool
+}
+
+// getLatestContainerStatus returns a KubernetesDeploymentStatus for the given component; this function blocks until it is available
+func getLatestContainerStatus(adapter *Adapter) *KubernetesDeploymentStatus {
```

Comment:
Unit tests for this function (and any others in the file) if possible

Reply:
I think `createStatusReconciler` is the best function in this file to unit test (the rest are either too simple, or themselves thoroughly tied into `createStatusReconciler`), so I've written unit tests for that. I think the rest are fine to be covered by the integration tests, let me know if you strongly disagree.

Reply:
👍